Add recipe for gnosis (γνῶσις)

thanosapollo opened the pull request:
### Brief summary of what the package does
Gnosis (γνῶσις), pronounced "noh-sis", *meaning knowledge in Greek*, is a spaced repetition system implementation for note taking and self testing.  [Read more](https://thanosapollo.org/user-manual/gnosis/)
### Direct link to the package repository

https://git.thanosapollo.org/gnosis/

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
  edit: I'm using a snapshot for emacsql, which is what the author recommends [here](https://github.com/magit/emacsql/issues/113), so there's this error using package-lint
      - `10:58: warning: Use a non-snapshot version number for dependency on "emacsql" if possible.`
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

progfolio commented:
A few notes:

Compiling on Emacs 30 signals the following warnings:

```
gnosis-algorithm.el:44:10: Warning: in defcustom for ‘gnosis-algorithm-interval’: ‘list’ without arguments
gnosis-algorithm.el:55:10: Warning: in defcustom for ‘gnosis-algorithm-ef’: ‘list’ without arguments
gnosis.el:339:2: Warning: docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')
```


```emacs-lisp
(defcustom gnosis-cloze-char "__"
  "Gnosis cloze character."
  :type 'string
  :group 'gnosis)
```

A better name for this option would be `gnosis-close-string` considering its type.


```emacs-lisp
  "Gnosis database file. WARNING: Do not change this value!")
```

You can drop the docstring warning since you're declaring the variable with `defconst`.


```emacs-lisp
(defmacro with-gnosis-buffer (&rest body)
```

You may want to reconsider the display design.
As it stands, each "display" function wraps its body with this macro.
It would be more efficient to do that once in the body of the functions which call all of those functions. e.g.

```emacs-lisp
(defun gnosis-review-mcq (id)
  "Display multiple choice answers for question ID."
  (with-current-buffer (get-buffer-create "*gnosis*")
    (gnosis-display--image id)
    (gnosis-display--question id)
    (let* ((choices (gnosis-get 'options 'notes `(= id ,id)))
       (answer (nth (- (gnosis-get 'answer 'notes `(= id ,id)) 1) choices))
       (user-choice (gnosis-mcq-answer id)))
      (if (string= answer user-choice)
          (progn (gnosis-review--update id 1)
             (message "Correct!"))
        (gnosis-review--update id 0)
        (message "False"))
      (gnosis-display--correct-answer-mcq answer user-choice)
      (gnosis-display--extra id)
      (gnosis-display--next-review id))))
      ```

This would remove the need for such a macro as well.
If you decide to keep it, the macro should still use the package's prefix. e.g. `gnosis-with-buffer`.
You should also add a debug declaration so edebug can step through the BODY:
https://www.gnu.org/software/emacs/manual/html_node/elisp/Instrumenting-Macro-Calls.html


The `gnosis-display--*` functions would be better named `gnosis--display-*`.


```emacs-lisp
        (cl-loop for note in notes
                 do (progn (gnosis-review-note note)
```

cl-loop's `do` clause is an implicit `progn`.


```emacs-lisp
";;\n;; You are editing a gnosis note. DO NOT change the value of id.\n\n")
```

If the ID should be protected, why not prevent editing it?
You could either display it as a propertized string with a non-nil read-only property, or not show it at all.


```emacs-lisp
(length= (emacsql gnosis-db [:select name :from sqlite-master :where (= type table)]) 6)
```

length= does not exist in Emacs < 28.1. If you intend to a minimum version of Emacs 27.2, you'll want to change that to `(eq (length ...`.

thanosapollo commented:
Nicholas Vollmer ***@***.***> writes:

> A few notes:
>
> Compiling on Emacs 30 signals the following warnings:
>
> gnosis-algorithm.el:44:10: Warning: in defcustom for ‘gnosis-algorithm-interval’: ‘list’ without arguments
> gnosis-algorithm.el:55:10: Warning: in defcustom for ‘gnosis-algorithm-ef’: ‘list’ without arguments
> gnosis.el:339:2: Warning: docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')
> (defcustom gnosis-cloze-char "__"
>   "Gnosis cloze character."
>   :type 'string
>   :group 'gnosis)
>

Fixed! (on the 'testing' branch)

> A better name for this option would be gnosis-close-string considering its type.
>

Agree!

> You can drop the docstring warning since you're declaring the variable with defconst.
>

Okay

> (defmacro with-gnosis-buffer (&rest body)
>
> You may want to reconsider the display design.
> As it stands, each "display" function wraps its body with this macro.
> It would be more efficient to do that once in the body of the functions which call all of those functions. e.g.
>
> (defun gnosis-review-mcq (id)
>   "Display multiple choice answers for question ID."
>   (with-current-buffer (get-buffer-create "*gnosis*")
>     (gnosis-display--image id)
>     (gnosis-display--question id)
>     (let* ((choices (gnosis-get 'options 'notes `(= id ,id)))
>        (answer (nth (- (gnosis-get 'answer 'notes `(= id ,id)) 1) choices))
>        (user-choice (gnosis-mcq-answer id)))
>       (if (string= answer user-choice)
>           (progn (gnosis-review--update id 1)
>              (message "Correct!"))
>         (gnosis-review--update id 0)
>         (message "False"))
>       (gnosis-display--correct-answer-mcq answer user-choice)
>       (gnosis-display--extra id)
>       (gnosis-display--next-review id))))
>       ```
>

Valid observation, I opted for using (with-current-buffer
(switch-to-buffer (get-buffer-create *gnosis*)) BODY) only on
`gnosis-review-note' instead. I think you will prefer this design as
well

```emacs-lisp
(defun gnosis-review-note (id)
  "Start review for note with value of id ID, if note is unsuspended."
  (cond ((gnosis-suspended-p id)
         (message "Note is suspended."))
        (t
	 (with-current-buffer (switch-to-buffer (get-buffer-create "*gnosis*"))
           (let ((type (gnosis-get 'type 'notes `(= id ,id))))
	     (gnosis-mode)
	     (pcase type
	       ("mcq" (gnosis-review-mcq id))
	       ("basic" (gnosis-review-basic id))
	       ("cloze" (gnosis-review-cloze id))
	       ("y-or-n" (gnosis-review-y-or-n id))
	       (_ (error "Malformed note type"))))))))
```               

I will see if I can rewrite this differently

> The `gnosis-display--*` functions would be better named `gnosis--display-*`.
>

Done.

>
> ```emacs-lisp
>         (cl-loop for note in notes
>                  do (progn (gnosis-review-note note)
>
> cl-loop's do clause is an implicit progn.
>

Thanks!

> ";;\n;; You are editing a gnosis note. DO NOT change the value of id.\n\n")
>
> If the ID should be protected, why not prevent editing it?
> You could either display it as a propertized string with a non-nil read-only property, or not show it at all.

True, I've changed this now so the id value will be inserted as
read-only


[...]
> length= does not exist in Emacs < 28.1. If you intend to a minimum version of Emacs 27.2, you'll want to change that to (eq (length ....

Done!


Thanks for taking a look at gnosis, I made the suggested changes & pushed them on
the 'testing' branch, with some additional changes on the algorithm
implementation.

I will be further testing them & merge them into master prob on Friday


-- 
Thanos Apollo
https://thanosapollo.org

riscy commented:
> Fixed! (on the 'testing' branch)

@thanosapollo I'll preferably only check your default branch for the updates, so let us know when those are ready.  On a quick glance I don't think they've been merged in yet.

And thank you again for deeper insights than I'd have had, @progfolio. :)

thanosapollo commented:
They've been merged on Thursday, but I've reverted some of them like ` length= does not exist in Emacs < 28.1. If you intend to a minimum version of Emacs 27.2, you'll want to change that to (eq (length ....` 

since I'm using compat, it's still  possible to use `length=` in emacs versions < 28.1, this was suggested by Philip Kaludercicon emacs-devel <https://lists.gnu.org/archive/html/emacs-devel/2024-01/msg00682.html> among others. 

I don't think the master branch has any issues, from a quick package-lint & compiling on 29.1 there's only:
  `10:58: warning: Use a non-snapshot version number for dependency on "emacsql" if possible.`. 

@riscy Feel free to skip adding this for now, but if you have time to check it out I would appreciate any feedback you can provide, thank you,

thanosapollo commented:
I've tried rewriting this with the built in sqlite support for emacs 29.1, but ended up reverting back to emacsql.

riscy commented:
Thanks!  Here's the latest --

First, do you intend to package gnosis-dev.el with this?  If not (it looks like a developer module on a very quick glance?) the recipe should exclude it, and you can disregard the nags below for that file --

`gnosis-algorithm.el` with package-lint 20240118.1725 and `package-lint-main-file` = "gnosis.el":
```
1 issue found:
10:0: error: Package-Requires outside the main file have no effect.
```

`gnosis-dev.el` with byte-compile using Emacs 29.1:
```
gnosis-dev.el:32:2: Error: No EmacSQL SQLite binary available: (Could not find C compiler, skipping SQLite build)
```
(I think this is coming from emacsql-sqlite, not your package, and it's because my env doesn't have a compiler to build sqlite; not your problem but fyi)

`gnosis-dev.el` with package-lint 20240118.1725 and `package-lint-main-file` = "gnosis.el":
```
1 issue found:
10:0: error: Package-Requires outside the main file have no effect.
```

`gnosis.el` with package-lint 20240118.1725 and `package-lint-main-file` = "gnosis.el":
```
1 issue found:
10:58: warning: Use a non-snapshot version number for dependency on "emacsql" if possible.
```
You can find the latest verson here: https://stable.melpa.org/#/emacsql

`gnosis.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- gnosis.el#L199: It's safer to sharp-quote function names; use `#'` (`#'max`)
- gnosis.el#L199: It's safer to sharp-quote function names; use `#'` (`#'length`)
- gnosis.el#L221: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L228: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L236: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L243: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L255: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L265: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L285: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L292: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L308: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L323: `concat` may be unneeded; `insert` can take multiple arguments
- gnosis.el#L1121: Can the `format` and `concat` be combined? 
   (This whole line could maybe just be `(insert ":" field " " (propertize value 'read-only t) "\n")`)
```

thanosapollo commented:
Chris Rayner ***@***.***> writes:

> Thanks! Here's the latest --
>
> First, do you intend to package gnosis-dev.el with this? If not (it looks like a developer module on a very quick glance?) the recipe should
> exclude it, and you can disregard the nags below for that file --

Thank you, I've updated the recipe.

>
> 1 issue found:
> 10:58: warning: Use a non-snapshot version number for dependency on "emacsql" if possible.
>
> You can find the latest verson here: https://stable.melpa.org/#/emacsql

I'm unsure of whether I should use the latest snapshot. Other packages
like org-roam rely on the "20230228" snapshot of emacsql, as discussed
in this issue:

<https://github.com/magit/emacsql/issues/113>.

-- 
Do not explain your philosophy.
Embody it.

Thanos Apollo
https://thanosapollo.org

riscy commented on recipes/gnosis:
```diff
@@ -0,0 +1,6 @@
+(gnosis :fetcher git
+        :url "https://git.thanosapollo.org/gnosis"
+        :files (:defaults
+                "*.el doc/*"
+                (:exclude
+                 "*-dev.el" "doc/*.org")))
```
```suggestion
(gnosis :fetcher git
        :url "https://git.thanosapollo.org/gnosis"
        :files (:defaults (:exclude "*-dev.el" "doc/*.org")))
```

riscy reviewed (commented):
> Other packages like org-roam rely on the "20230228" snapshot of emacsql

This is not a blocker - as long as you understand the issue and have made a choice.


I have suggested an alternative recipe that yields the same files (all you really needed was `:defaults`).  Let me know if this looks good to you and we can merge.

thanosapollo commented:
Thank you, this looks good :)

riscy commented:
This popped up in a different context, but in the intersest of keeping things moving I will merge --

`gnosis.el` with byte-compile using Emacs 29.1:
```
gnosis.el:72:2: Warning: custom-declare-variable `gnosis-auto-push-command' docstring has wrong usage of unescaped single quotes (use \= or different quoting)
```

Thanks again for this.
